### PR TITLE
Minimax player

### PR DESCRIPTION
**Pull request by gemcfadyen**
@jsuchy @ecomba

This PR has the minimax player integrated into the TTT Game.

The following is outstanding from this PR:
1) Alpha Beta pruning - as such, if the player chooses the option `Ai vs Human`, the first move takes too long (about 12 seconds)
2) Unbeatable test to prove that the ai player can not loose a game
3) Inheritance amongst the players.

### CODE REVIEW

File: lib/ai_player.rb [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    if is_max_player && (result_score >= best_score_so_far.get_score)
+      return ScoredMove.new(move, result_score)
+    elsif !is_max_player && (result_score < best_score_so_far.get_score)
+      return ScoredMove.new(move, result_score)
+    end
+    return best_score_so_far
+  end
+end
+
+class ScoredMove
```
Comment (jsuchy) [inline lib/ai_player.rb:87]: replace with just a hash?

File: lib/ai_player.rb [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    @score = score 
+  end
+
+  private
+
+  attr_accessor :move
+  attr_accessor :score
+end
+
+class MaximisingPlayerInitialScore < ScoredMove
```
Comment (jsuchy) [inline lib/ai_player.rb:108]: consider using negamax to simplify. can probably get rid of all of these classes. applying 4 rules of simple design, minimise the number of classes.

File: lib/command_line_ui.rb [hunk 3]
```diff
@@ -29,33 +29,55 @@ def print_game_status(board)
...
+  end
+
   private
 
   attr_reader :writer, :reader
 
   def get_user_input_for_move
     writer.ask_for_next_move
-    reader.get_input  
+    value = reader.get_input
```
Comment (jsuchy) [inline lib/command_line_ui.rb:43]: avoid temp variables unless they actually add value...

File: lib/command_line_ui.rb [hunk 4]
```diff
@@ -29,33 +29,55 @@ def print_game_status(board)
...
 
   def get_user_input_for_move
     writer.ask_for_next_move
-    reader.get_input  
+    value = reader.get_input  
+  end
+
+  def get_user_input_for_player_options
+    writer.show_player_options 
+    value = reader.get_input
```
Comment (jsuchy) [inline lib/command_line_ui.rb:48]: no need to store in a temp variable

**Comment by gemcfadyen** [comment]
@jsuchy I am merging this PR now that coverage is 100% (there were two unused methods that were knocking it down to 99%). I will address the comments you made on minimax in a separate PR.

**Pull request merged**